### PR TITLE
Handle exceptions in startupService

### PR DESCRIPTION
**Pull request by brandicorel**

### CODE REVIEW

File: src/AspNetCore/Health/StartupHealthChecker.cs [hunk 1]
```diff
@@ -23,9 +21,11 @@ public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, Canc
             var data = new ReadOnlyDictionary<string, object>(
                 this.Services.ToDictionary(
                     s => s.GetType().Name,
-                    s => (object) (s.HasCompleted ? "Completed" : "Running")));
+                    s => (object) (
+                                      !s.HasCompleted.HasValue ? "Running" :
+                                      s.HasCompleted.Value     ? "Completed" : "Failed")));
 
-            var healthCheckResult = this.Services.Any(s => !s.HasCompleted)
+            var healthCheckResult = this.Services.Any(s => s.HasCompleted != true)
```
Comment (KevinRecuerda) [inline src/AspNetCore/Health/StartupHealthChecker.cs:28]: maybe more explicit
```csharp
            var healthCheckResult = this.Services.All(s => s.HasCompleted == true)
                                        ? HealthCheckResult.Healthy(data: data)
                                        : HealthCheckResult.Unhealthy(data: data);
```

File: src/AspNetCore/Health/StartupHostedService.cs [hunk 2]
```diff
@@ -3,21 +3,38 @@
...
         protected override async Task ExecuteAsync(CancellationToken stoppingToken)
         {
-            await this.InnerExecute(stoppingToken);
+            await Task.Run(() => this.InnerExecute(stoppingToken), stoppingToken)
+                      .ContinueWith(
+                           t =>
+                           {
+                               this.HasCompleted = t.IsCompletedSuccessfully;
+
+                               var type = this.GetType().Name;
```
Comment (KevinRecuerda) [inline src/AspNetCore/Health/StartupHostedService.cs:31]: should we add property for this? (which could be virtual)

File: src/AspNetCore/Health/StartupHostedService.cs [hunk 3]
```diff
@@ -3,21 +3,38 @@
...
     public interface IStartupHostedService : IHostedService
     {
-        public bool HasCompleted { get; set; }
+        public bool? HasCompleted { get; set; }
     }
 
     public abstract class StartupHostedService : BackgroundService, IStartupHostedService
     {
-        public bool HasCompleted { get; set; }
+        public StartupHostedService(ILogger logger)
```
Comment (brandicorel) [inline src/AspNetCore/Health/StartupHostedService.cs:15]: Is it standard to add a ILogger with DI, or should I allow null and manage when no logger are specified?

File: src/AspNetCore/Health/StartupHealthChecker.cs [hunk 4]
```diff
@@ -23,9 +21,11 @@ public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, Canc
             var data = new ReadOnlyDictionary<string, object>(
                 this.Services.ToDictionary(
                     s => s.GetType().Name,
-                    s => (object) (s.HasCompleted ? "Completed" : "Running")));
+                    s => (object) (
+                                      !s.HasCompleted.HasValue ? "Running" :
+                                      s.HasCompleted.Value     ? "Completed" : "Failed")));
```
Comment (KevinRecuerda) [inline src/AspNetCore/Health/StartupHealthChecker.cs:26]: maybe add extensions method for this:
`ToString(this bool? x, string trueValue, string falseValue, string nullValue)`

File: src/AspNetCore/Health/StartupHostedService.cs (hunk 3 above)
Comment (KevinRecuerda) [reply src/AspNetCore/Health/StartupHostedService.cs:15]: must be mandatory

File: src/AspNetCore/Health/StartupHostedService.cs [hunk 5]
```diff
@@ -3,21 +3,38 @@
...
+        public bool?   HasCompleted { get; set; }
 
         protected override async Task ExecuteAsync(CancellationToken stoppingToken)
         {
-            await this.InnerExecute(stoppingToken);
+            await Task.Run(() => this.InnerExecute(stoppingToken), stoppingToken)
+                      .ContinueWith(
+                           t =>
+                           {
+                               this.HasCompleted = t.IsCompletedSuccessfully;
```
Comment (brandicorel) [inline src/AspNetCore/Health/StartupHostedService.cs:29]: LogInfo(service completed) ?

**Review by KevinRecuerda — APPROVED** [review]

**Pull request merged**